app/models/miq_server/worker_management/monitor/validation.rb

```diff
@@ -4,22 +4,50 @@ module MiqServer::WorkerManagement::Monitor::Validation
   def validate_worker(w)
     return true if MiqEnvironment::Command.is_podified?
 
-    time_threshold   = get_time_threshold(w)
-    memory_threshold = get_memory_threshold(w)
+    if !worker_heartbeat_valid?(w)
+      stop_worker(w, MiqServer::NOT_RESPONDING)
+      return false
+    end
 
-    w.validate_active_messages
+    if !worker_memory_valid?(w)
+      stop_worker(w)
+      return false
+    end
 
-    validate_heartbeat(w)
+    return true unless worker_get_monitor_status(w.pid).nil?
+
+    true
+  end
+
+  def validate_active_messages(processed_worker_ids = [])
+    actives = MiqQueue.where(:state => 'dequeue').includes(:handler)
+    actives.each do |msg|
+      next if processed_worker_ids.include?(msg.handler_id)
+
+      # Exclude messages on starting/started servers
+      handler = msg.handler
+      handler_server = handler.respond_to?(:miq_server) ? handler.miq_server : handler
+      if handler_server.kind_of?(MiqServer) && handler_server != self
+        next if [MiqServer::STATUS_STARTED, MiqServer::STATUS_STARTING].include?(handler_server.status)
+      end
+
+      msg.check_for_timeout(_log.prefix)
+    end
+  end
 
+  def worker_heartbeat_valid?(w)
+    time_threshold = get_time_threshold(w)
     if time_threshold.seconds.ago.utc > w.last_heartbeat
       msg = "#{w.format_full_log_msg} has not responded in #{Time.now.utc - w.last_heartbeat} seconds, restarting worker"
       _log.error(msg)
```

Comment:
It's weird that the log message is here but it doesn't actually restart it...I think we should move the log message up, or create a new method called something like `handle_worker_on_invalid_heartbeat` that does the "actual work"

Reply:
Yeah, I was struggling with this one because in the memory case we actually use the threshold value in the log message and event (which we probably should be doing here as well), but I figured the important part was seeing that we stop the worker if it fails one of the checks and the logging and event were more acceptable side-effects.

Reply:
I still don't like that these predicate methods have a side effect of logging, but we can change this in a follow up as I know a lot will change in this area moving forward.